src/doorman/Doorman.cpp

```diff
@@ -0,0 +1,126 @@
+    listenSock_ = Fd::socket(PF_INET, SOCK_STREAM, IPPROTO_TCP);
+
+    sockaddr_in sin{};
+    memset(&sin, 0, sizeof(sin));
+    sin.sin_family = PF_INET;
+    sin.sin_port = htons(port);
+    sin.sin_addr.s_addr = htonl(INADDR_ANY);
+    listenSock_.setsockopt(SOL_SOCKET, SO_REUSEADDR, static_cast<int>(1))
+        .setsockopt(SOL_SOCKET, SO_LINGER, linger{true, 2})
+        .bind(sin)
+        .listen(4);
+
+    log_.info("Doorman is ready to rock on port {}", port);
+}
+
+void Doorman::poll() {
+    mud_.poll();
+    socket_poll();
+    remove_dead_channels();
+}
+
+void Doorman::remove_dead_channels() {
+    // As channels can get deleted while we're iterating over the master list of channels, we protect ourselves from the
+    // "delete this" problem by deferring their actual removal until we know that no calls to the Channel are on the
+    // call stack.
+    for (auto id : channels_to_remove_) {
+        log_.debug("Removed channel {} from master channel list", id);
+        channels_.erase(id);
+    }
+    channels_to_remove_.clear();
+}
+
+void Doorman::socket_poll() {
+    fd_set input_fds, exception_fds;
+    FD_ZERO(&input_fds);
+    FD_ZERO(&exception_fds);
+    FD_SET(listenSock_.number(), &input_fds);
+    int max_fd = listenSock_.number();
+    if (mud_.fd_ok()) {
+        FD_SET(mud_.fd().number(), &input_fds);
+        FD_SET(mud_.fd().number(), &exception_fds);
+        max_fd = std::max(max_fd, mud_.fd().number());
+    }
+    for_each_channel([&](Channel &channel) { max_fd = std::max(channel.set_fds(input_fds, exception_fds), max_fd); });
+
+    timeval timeOut = {1, 0}; // Wakes up once a second to do housekeeping
+    int nFDs = select(max_fd + 1, &input_fds, nullptr, &exception_fds, &timeOut);
+    if (nFDs == -1 && errno != EINTR) {
+        log_.error("Unable to select()!");
+        perror("select");
+        exit(1);
+    }
+    if (nFDs <= 0)
+        return;
+
+    // Mud handling.
+    if (FD_ISSET(mud_.fd().number(), &exception_fds)) {
+        log_.warn("Got exception from MUD file descriptor");
+        mud_.close();
+    } else if (FD_ISSET(mud_.fd().number(), &input_fds))
+        mud_.process_mud_message();
+
+    if (FD_ISSET(listenSock_.number(), &input_fds))
+        accept_new_connection();
+
+    for_each_channel([&](Channel &channel) { channel.check_fds(input_fds, exception_fds); });
+}
+
+void Doorman::accept_new_connection() {
+    sockaddr_in incoming{};
+    socklen_t len = sizeof(incoming);
+    try {
+        auto newFd = listenSock_.accept(reinterpret_cast<sockaddr *>(&incoming), &len);
+
+        if (channels_.size() == MaxChannels) {
+            newFd.write("Xania is out of channels!\n\rTry again soon\n\r"sv);
+            log_.warn("Rejected connection - out of channels");
+            return;
+        }
+
+        auto id = id_allocator_.reserve();
+        auto insert_rec = channels_.try_emplace(id->id(), *this, mud_, id, std::move(newFd), incoming);
```

Comment:
One style comment:  I was trying to track down where we were instantiating Channels. And by hiding it in a try_emplace() it makes it a bit tricky to do that :)  Probably the same for any other collections we're writing to with an "inlined" construction rather than a local var.

Reply:
Agreed: sometimes we're stuck with this as one can't move the object one creates: that's the case here. We can't make a temporary Channel and then put it in the map, we must construct it "in place" inside the map. That's a side effect of the fact the channels' address must remain consistent as it is passed as a reference around (particularly as a param to its internal telnet handler), else we could "move" it. `try_emplace` was added in C++17 specifically for this case.

An alternative is to use a `unique_ptr` to a Channel and store that in the map: then we only ever need to create it once in a known location, which is an alternative. Slightly higher overhead (a pointer chase) and minor cognitive load (`->` instead of `.`). Given this caused a bit of confusion, I'll rephrase that way, as we don't expose the pointer-ness or value-ness of the Channel elsewhere (yay encapsulation).